voyager-navigator/src/commonMain/kotlin/cafe/adriel/voyager/navigator/model/NavigatorScreenModelStore.kt

```diff
@@ -0,0 +1,33 @@
+package cafe.adriel.voyager.navigator.model
+
+import cafe.adriel.voyager.core.concurrent.ThreadSafeMap
+import cafe.adriel.voyager.core.model.ScreenModel
+import cafe.adriel.voyager.navigator.Navigator
+import cafe.adriel.voyager.navigator.lifecycle.NavigatorDisposable
+
+public class NavigatorScreenModelStore : NavigatorDisposable {
+    private val screenModels: MutableMap<String, ScreenModel> = ThreadSafeMap()
+
+    override fun onDispose(navigator: Navigator) {
+        screenModels.forEach { entry ->
+            entry.value.onDispose()
+        }
+        screenModels.clear()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    public fun <T : ScreenModel> get(key: String): T? = screenModels[key] as? T
+
+    @Suppress("UNCHECKED_CAST")
+    public fun <T : ScreenModel> getOrPut(
+        key: String,
+        factory: () -> T
+    ): T {
+        var model = screenModels[key] as? T
+        if (model == null) {
+            model = factory()
+            screenModels[key] = model
+        }
+        return model
+    }
+}
```

Comment:
I think we will have a issue with `dependencies` from ScreenModel, for example `screenModelScope` will be closed when the first Screen that is holding a shared ScreenModel leaves the Stack.